File: app-android/src/main/java/com/ustadmobile/port/android/netwokmanager/BleEntryStatusTaskAndroid.java
package com.ustadmobile.port.android.netwokmanager;

import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothManager;
import android.content.Context;
import android.os.Build;
import android.support.annotation.RequiresApi;

import com.ustadmobile.core.impl.UMLog;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;
import com.ustadmobile.lib.db.entities.NetworkNode;
import com.ustadmobile.port.sharedse.networkmanager.BleEntryStatusTask;
import com.ustadmobile.port.sharedse.networkmanager.BleMessage;
import com.ustadmobile.port.sharedse.networkmanager.BleMessageResponseListener;
import com.ustadmobile.port.sharedse.networkmanager.BleMessageUtil;
import com.ustadmobile.port.sharedse.networkmanager.NetworkManagerBle;

import java.util.List;

import static com.ustadmobile.port.sharedse.networkmanager.NetworkManagerBle.ENTRY_STATUS_REQUEST;

/**
 * This class handles all android specific entry status check from a peer device also,
 * it is responsible for creating BLE GATT client callback.
 *
 * <b>Note: Operation Flow</b>
 * <p>
 * - Once {@link BleEntryStatusTaskAndroid#run()} is called, it creates
 * {@link BleMessageGattClientCallback} and pass the list of entries to be checked
 * and peer device to be checked from. After entry status check
 * {@link BleEntryStatusTask#onResponseReceived} will be called to report back the results.
 *
 * <p>
 * Use {@link BleEntryStatusTaskAndroid#run()} to start executing the task itself,
 * this method will be called in {@link NetworkManagerBle#startMonitoringAvailability}
 * when pending task to be executed is found.
 *
 *
 * @see BleMessageGattClientCallback
 * @see BleEntryStatusTask
 * @see NetworkManagerBle
 *
 *  @author kileha3
 */
@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
public class    BleEntryStatusTaskAndroid extends BleEntryStatusTask {

    private BleMessageGattClientCallback mCallback;

    private BluetoothManager bluetoothManager;

    private NetworkNode peerToCheck;

    private Context context;

    /**
     * Constructor to be used when creating platform specific instance of BleEntryStatusTask
     * @param context Platform specific application context.
     * @param entryUidsToCheck List of Id's to be checked for availability from a peer device.
     * @param peerToCheck Peer device for those entries to be checked from.
     */
    public BleEntryStatusTaskAndroid(Context context, List<Long> entryUidsToCheck,
                              NetworkNode peerToCheck) {
        super(context,entryUidsToCheck,peerToCheck);
        this.context = context;
        this.peerToCheck = peerToCheck;
        byte [] messagePayload = BleMessageUtil.bleMessageLongToBytes(entryUidsToCheck);
        this.message = new BleMessage(ENTRY_STATUS_REQUEST,messagePayload);
    }

    /**
     * Constructor to be used when creating platform specific instance of BleEntryStatusTask
     * responsible for sending custom messages apart from entry status check.
     * @param context Platform specific application context.
     * @param message Message to be sent
     * @param peerToSendMessageTo peer to send message to
     * @param responseListener Message response listener object
     */
    public BleEntryStatusTaskAndroid(Context context , BleMessage message,
                                     NetworkNode peerToSendMessageTo,
                                     BleMessageResponseListener responseListener){
        super(context,message,peerToSendMessageTo, responseListener);
    }


    /**
     * Set bluetooth manager for BLE GATT communication
     * @param bluetoothManager BluetoothManager instance
     */
    void setBluetoothManager(BluetoothManager bluetoothManager){
        this.bluetoothManager = bluetoothManager;
    }

    /**
     * Start entry status check task
     */
    @Override
    public void run() {
       try{
           mCallback = new BleMessageGattClientCallback(message);
           mCallback.setOnResponseReceived(this);
           BluetoothDevice destinationPeer = bluetoothManager.getAdapter()
                   .getRemoteDevice(peerToCheck.getBluetoothMacAddress());
            BluetoothGatt gatt= destinationPeer.connectGatt(context,false,mCallback);
            if(gatt == null){
                UstadMobileSystemImpl.l(UMLog.ERROR,695,
                        "Failed to connect to "+destinationPeer.getAddress());
            }else{
                UstadMobileSystemImpl.l(UMLog.DEBUG,695,
                        "Connecting to "+destinationPeer.getAddress());
            }
       }catch (IllegalArgumentException e){
           UstadMobileSystemImpl.l(UMLog.ERROR,695,
                   "Wrong address format provided",e);
       }
    }

    /**
     * Get BleMessageGattClientCallback instance
     * @return Instance of a BleMessageGattClientCallback
     */
    BleMessageGattClientCallback getGattClientCallback() {
        return mCallback;
    }
}

File: sharedse/src/androidMain/kotlin/com/ustadmobile/sharedse/network/BleEntryStatusTaskAndroid.kt
package com.ustadmobile.sharedse.network

import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothManager
import android.content.Context
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.annotation.VisibleForTesting
import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.impl.UMLog
import com.ustadmobile.lib.db.entities.NetworkNode
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.io.IOException

/**
 * This class handles all android specific entry status check from a peer device also,
 * it is responsible for creating BLE GATT client callback.
 *
 * **Note: Operation Flow**
 *
 *
 * - Once [BleEntryStatusTaskAndroid.run] is called, it creates
 * [BleMessageGattClientCallback] and pass the list of entries to be checked
 * and peer device to be checked from. After entry status check
 * [BleEntryStatusTask.onResponseReceived] will be called to report back the results.
 *
 *
 *
 * Use [BleEntryStatusTaskAndroid.run] to start executing the task itself,
 * this method will be called in [NetworkManagerBle.startMonitoringAvailability]
 * when pending task to be executed is found.
 *
 *
 * @see BleMessageGattClientCallback
 *
 * @see BleEntryStatusTask
 *
 * @see NetworkManagerBle
 *
 *
 * @author kileha3
 */
@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
class BleEntryStatusTaskAndroid : BleEntryStatusTask {

    private val gattClientCallbackManager: GattClientCallbackManager

    private var bluetoothManager: BluetoothManager? = null

    private var mGattClient: BluetoothGatt? = null

    private var managerBle: NetworkManagerBle? = null

    /**
     * Constructor to be used when creating platform specific instance of BleEntryStatusTask
     * @param context Platform specific application context.
     * @param entryUidsToCheck List of Id's to be checked for availability from a peer device.
     * @param peerToCheck Peer device for those entries to be checked from.
     */
    constructor(gattClientCallbackManager: GattClientCallbackManager, context: Context, managerAndroidBle: NetworkManagerBle,
                entryUidsToCheck: List<Long>, peerToCheck: NetworkNode)
            : super(context, managerAndroidBle, entryUidsToCheck, peerToCheck) {
        this.gattClientCallbackManager = gattClientCallbackManager
        this.managerBle = managerAndroidBle
        this.context = context
        val messagePayload = BleMessageUtil.bleMessageLongToBytes(entryUidsToCheck)
        this.message = BleMessage(NetworkManagerBleCommon.ENTRY_STATUS_REQUEST,
                BleMessage.getNextMessageIdForReceiver(peerToCheck.bluetoothMacAddress!!),
                messagePayload)
    }

    /**
     * Constructor to be used when creating platform specific instance of BleEntryStatusTask
     * responsible for sending custom messages apart from entry status check.
     * @param context Platform specific application context.
     * @param message Message to be sent
     * @param peerToSendMessageTo peer to send message to
     * @param responseListener Message response listener object
     */
    constructor(gattClientCallbackManager: GattClientCallbackManager,
                context: Context, managerBle: NetworkManagerBle, message: BleMessage,
                peerToSendMessageTo: NetworkNode,
                responseListener: BleMessageResponseListener)
            : super(context, managerBle, message, peerToSendMessageTo, responseListener) {
        this.gattClientCallbackManager = gattClientCallbackManager
    }


    /**
     * Set bluetooth manager for BLE GATT communication
     * @param bluetoothManager BluetoothManager instance
     */
    @VisibleForTesting
    fun setBluetoothManager(bluetoothManager: BluetoothManager) {
        this.bluetoothManager = bluetoothManager
    }


    /**
     * Start entry status check task
     */
    override fun sendRequest() {
        try {
            //TODO: this changes to get the clientcallback from a pool
            val bluetoothAddr = networkNode.bluetoothMacAddress
            if(bluetoothAddr == null) {
                //Illegal Argument - should never get here
                throw IllegalArgumentException("SendRequest called with null bluetooth destination")
            }

            val gattClientCallback = gattClientCallbackManager.getGattClient(bluetoothAddr)
            GlobalScope.launch {
                try {
                    val messageReceived = gattClientCallback.sendMessage(message)
                    onResponseReceived(bluetoothAddr, messageReceived, null)
                }catch(e: Exception) {
                    onResponseReceived(bluetoothAddr, null, e)
                }
            }

        } catch (e: IllegalArgumentException) {
            UMLog.l(UMLog.ERROR, 698,
                    "Wrong address format provided", e)
        }

    }
}
